Census ap data import

drewsamnick opened the pull request:
This is the code to import the census AP data from files in S3. For now I am not having this part of the standard seed tasks (seed:all, etc.) because we found some issues with the school code data that we need to work out. 

I debated wether it made sense to move this code out of seed.rake and if so where it would make sense to put it. It doesn't feel like it logically is part of the model and that it is more properly part of the DB seeding logic. I'm happy to move it someplace else though.

drewsamnick commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create table for AP data](https://trello.com/c/dk3MDBPH/77-create-table-for-ap-data)

aoby commented:
A couple concerns:

1. How large are the files / how long does the seeding take? Should we have a stubbed version to save time on dev environments like we do for schools and districts?

2. What happens for, say outside contributors, who don't have S3 access? Again, perhaps we should have local stubs.

davidsbailey commented on dashboard/app/models/census/ap_cs_offering.rb:
```diff
@@ -26,5 +26,60 @@ class Census::ApCsOffering < ApplicationRecord
+  end
+
+  def self.seed
+    if CDO.stub_school_data
```
Nice reuse of existing config 👍

drewsamnick commented:
Ok, I made several changes. 
1. Added stub data. Since the AP data depends on the schools data I used the same config flag to control wether to use stub data.
2. Moved the seed logic into the models.
3. Added a way to track which S3 object versions have already been seeded and skip them. This will allow us to seed once but still reseed if the file is updated.

aoby commented on dashboard/app/models/census/ap_school_code.rb:
```diff
@@ -19,4 +19,52 @@ class Census::ApSchoolCode < ApplicationRecord
+
+  def self.seed_from_s3
+    etag = AWS::S3.create_client.head_object({bucket: CENSUS_BUCKET_NAME, key: CSV_OBJECT_KEY}).etag
+    unless SeededS3Object.where(bucket: CENSUS_BUCKET_NAME, key: CSV_OBJECT_KEY, etag: etag).count > 0
```
use `exists?` rather than `count > 0`

drewsamnick replied:
Done.

aoby commented on dashboard/db/migrate/20180110233855_create_seeded_s3_objects.rb:
```diff
@@ -0,0 +1,13 @@
+      t.timestamps
+    end
+
+    add_index :seeded_s3_objects, [:bucket, :key, :etag]
```
Should we also have an index on just `[:bucket, :key]` without `:etag`? I can see wanting a quick check to know whether a particular S3 file is seeded at all (regardless of contents / version).

drewsamnick replied:
We shouldn't need an additional index. Since bucket and key are the leading edge of the index it should be used for queries even if they don't include an etag. 

```
mysql> explain select * from seeded_s3_objects o where o.bucket='a bucket' and o.key='a key';
explain select * from seeded_s3_objects o where o.bucket='a bucket' and o.key='a key';
+----+-------------+-------+------------+------+----------------------------------------------------+----------------------------------------------------+---------+-------------+------+----------+-------+
| id | select_type | table | partitions | type | possible_keys                                      | key                                                | key_len | ref         | rows | filtered | Extra |
+----+-------------+-------+------------+------+----------------------------------------------------+----------------------------------------------------+---------+-------------+------+----------+-------+
|  1 | SIMPLE      | o     | NULL       | ref  | index_seeded_s3_objects_on_bucket_and_key_and_etag | index_seeded_s3_objects_on_bucket_and_key_and_etag | 1536    | const,const |    1 |   100.00 | NULL  |
+----+-------------+-------+------------+------+----------------------------------------------------+----------------------------------------------------+---------+-------------+------+----------+-------+
1 row in set, 1 warning (0.01 sec)
```

aoby commented on dashboard/app/models/census/ap_cs_offering.rb:
```diff
@@ -26,5 +26,70 @@ class Census::ApCsOffering < ApplicationRecord
+        object_key = "ap_cs_offerings/#{course}-#{school_year}-#{school_year + 1}.csv"
+        begin
+          etag = AWS::S3.create_client.head_object({bucket: CENSUS_BUCKET_NAME, key: object_key}).etag
+          unless SeededS3Object.where(bucket: CENSUS_BUCKET_NAME, key: object_key, etag: etag).count > 0
```
s/`count > 0`/`exists?`

drewsamnick replied:
Done.
